server/src/main/java/org/elasticsearch/index/mapper/DocCountFieldMapper.java

```diff
@@ -0,0 +1,160 @@
+            FIELD_TYPE.setIndexOptions(IndexOptions.NONE);
+            FIELD_TYPE.freeze();
+        }
+    }
+
+    public static class Builder extends FieldMapper.Builder<DocCountFieldMapper.Builder> {
+
+        public Builder(String name) {
+            super(name, Defaults.FIELD_TYPE);
+            builder = this;
+        }
+
+        @Override
+        public DocCountFieldMapper build(BuilderContext context) {
+            DocCountFieldType defaultFieldType = new DocCountFieldType(buildFullName(context), hasDocValues, meta);
+            return new DocCountFieldMapper(name, fieldType, defaultFieldType);
+        }
+    }
+
+    public static class TypeParser implements Mapper.TypeParser {
+
+        public TypeParser() {
+        }
+
+        @Override
+        public DocCountFieldMapper.Builder parse(String name, Map<String, Object> node, ParserContext parserContext)
+            throws MapperParsingException {
+            DocCountFieldMapper.Builder builder = new DocCountFieldMapper.Builder(name);
+            parseField(builder, name, node, parserContext);
+            return builder;
+        }
+    }
+
+    public static final class DocCountFieldType extends MappedFieldType {
+
+        public DocCountFieldType(String name, boolean hasDocValues, Map<String, String> meta) {
+            super(name, false, hasDocValues, TextSearchInfo.NONE, meta);
+        }
+
+        public DocCountFieldType(String name) {
+            this(name, true, Collections.emptyMap());
+        }
+
+        @Override
+        public String typeName() {
+            return CONTENT_TYPE;
+        }
+
+        @Override
+        public Query existsQuery(QueryShardContext context) {
+            return new DocValuesFieldExistsQuery(name());
+        }
+
+        @Override
+        public Query termQuery(Object value, QueryShardContext context) {
+            throw new QueryShardException(context, "Field [" + name() + " ]of type [" + CONTENT_TYPE + "] is not searchable");
+        }
+    }
+
+    protected DocCountFieldMapper(
+        String simpleName,
+        FieldType fieldType,
+        MappedFieldType defaultFieldType) {
+        super(simpleName, fieldType, defaultFieldType, MultiFields.empty(), CopyTo.empty());
+    }
+
+    @Override
+    protected void parseCreateField(ParseContext context) throws IOException {
+        Number value;
+        if (context.parser().currentToken() == XContentParser.Token.VALUE_NUMBER) {
+            value = context.parser().numberValue().floatValue();
+        } else {
+            return;
+        }
+
+        if (value != null) {
+            if (value.longValue() <= 0 || value.floatValue() != value.longValue()) {
+                throw new IllegalArgumentException(
+                    "Field [" + fieldType().name() + "] must be a positive integer.");
+            }
+
+            final Field docCount = new NumericDocValuesField(name(), value.longValue());
```

Comment:
I wonder if you can use a canonical name for the Lucene field since we allow a single doc_count field per mapping. Something like ` _doc_count` would be easier to manipulate rather than looking in the mapping if a doc_count field exists for every request ?